rumqttc/src/framed.rs

```diff
@@ -23,28 +25,55 @@ impl Network {
         max_incoming_size: usize,
         max_outgoing_size: usize,
         timeout: Duration,
+        buffer_capacity: usize,
     ) -> Network {
         let socket = Box::new(socket) as Box<dyn AsyncReadWrite>;
         let codec = Codec {
             max_incoming_size,
             max_outgoing_size,
         };
-        let framed = Framed::new(socket, codec);
+        let framed = Framed::with_capacity(socket, codec, buffer_capacity);
 
-        Network { framed, timeout }
+        Network {
+            framed,
+            timeout,
+            buffer_capacity,
+        }
     }
 
     pub async fn read(&mut self) -> Result<Incoming, StateError> {
         match self.framed.next().await {
             Some(Ok(packet)) => Ok(packet),
-            Some(Err(mqttbytes::Error::InsufficientBytes(_))) | None => unreachable!(),
+            Some(Err(mqttbytes::Error::InsufficientBytes(_))) => unreachable!(),
             Some(Err(e)) => Err(StateError::Deserialization(e)),
+            None => Err(StateError::ConnectionClosed),
         }
     }
 
-    pub async fn send(&mut self, packet: Packet) -> Result<(), crate::state::StateError> {
-        match timeout(self.timeout, self.framed.send(packet)).await {
-            Ok(inner) => inner.map_err(Into::into),
+    /// Write packets into buffer, flushes `Connect`/`PingReq`/`PingResp` packets instantly,
+    /// or on breaching buffer capacity
+    pub async fn write(&mut self, packet: Packet) -> Result<(), StateError> {
+        let packet_size = packet.size();
+        let should_flush = match packet {
+            Packet::Connect(_) | Packet::PingReq | Packet::PingResp => true,
+            _ => false,
+        };
+        self.framed
+            .feed(packet)
+            .await
+            .map_err(StateError::Deserialization)?;
+
+        if should_flush || self.framed.write_buffer().len() + packet_size >= self.buffer_capacity {
```

Comment:
`FramedWrite` does that job for you [here](https://docs.rs/tokio-util/latest/src/tokio_util/codec/framed_impl.rs.html#264). I think removing the manual buffer size check is good and avoid not necessary complexity here. 

I'd also do the "when to flush" logic in `EventLoop` and not in here because it must reflect MQTT and the IO part should be agnostic.

Add a loop (up to e.g 10) around the `select!` that breaks if `pending` or `request_rx` is empty and the rx buffer of `Framed` is empty after a `fill_buf`. There is no need for extra logic in `Network`.

like this:

```
for _ in 0..10 {
  select! {
    r = next_request() => {
       packet = state.process(r)
       network.feed(packet)
    }
    p = network.read() => {
       packet = state.handle_incoming(p)
       network.feed(packet)
       if matches(packet, PingResp) {
           // Ensure the flush by breaking.
           break;
       }
    }
    _ = ping_interval.tick() => {
       packet = PingReq
       network.feed(packet)
    }
  }

  let data_pending = network.fill_rx_buf(); // get inner of framed; get read_buffer_mut, fill it.

  // Break if there is nothing to do. 
  if pending.is_empty() && requests_rx.is_empty() && !data_pending() {
    break;
  }
}
network.flush()
```